### PR TITLE
feat(provenance): persist and resolve kernel provenance artifacts

### LINKED ISSUE #2578: Emit spyre_provenance.json sidecar artifact
## Overview

Generate a `spyre_provenance.json` sidecar file during compilation that maps each `debug_handle` through every IR stage. This artifact enables offline analysis tools and the visualization viewer to reconstruct the full provenance chain without re-running compilation.

## Acceptance Criteria

- [ ] Define the JSON schema for `spyre_provenance.json`
- [ ] Each entry keyed by `debug_handle.id` maps through:
  - Source frame (file, line, col, function name)
  - Pre-grad FX node name
  - Post-grad FX node name
  - LoopLevelIR op identifier
  - OpSpec identifier
  - SDSC kernel name
  - Profiler event name/ID
- [ ] Schema mirrors upstream PyTorch's `inductor_provenance_tracking_node_mappings.json` structure for consistency
- [ ] Implement the generator: collect mappings during compilation and serialize at the end
- [ ] Output path configurable via environment variable (default: alongside compiled artifacts)
- [ ] Handle fusion correctly: fused entries reference all constituent source handles
- [ ] Unit tests for schema validation and round-trip
- [ ] Integration test: compile a model, verify JSON output contains correct mappings

## Schema Design (Draft)

```json
{
  "version": "1.0",
  "model": "granite-3b",
  "handles": {
    "42": {
      "source": {"file": "model.py", "line": 117, "col": 8, "function": "forward"},
      "aten_op": "aten.linear",
      "pre_grad_fx": "linear_0",
      "post_grad_fx": "fn_0_linear",
      "loop_level_ir": "llir_op_3",
      "op_spec": "opspec_7",
      "sdsc_kernel": "kernel_compute_0",
      "profiler_event": "spyre_kernel_42_linear_at_model_py_117",
      "fused_from": null
    }
  }
}
```

## Key Files

- New file: `torch_spyre/_inductor/provenance.py` (or extend from #2576)
- `torch_spyre/_inductor/codegen/compute_ops.py` (collection point during codegen)
- `torch_spyre/_inductor/scheduler.py` (final assembly of kernel ↔ handle mapping)

## Timeline

July 24 – July 31, 2026 (second half of Phase 3)

## Dependencies

- debug_handle schema (#2575)
- Profiler annotation (#2577) — profiler event names needed for the mapping

### PR DESCRIPTION
**Pull request by gsmoon97**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Adds a durable mapping from Spyre profiler events to source provenance.

- Publishes a strict, versioned `spyre_provenance.json` v1 sidecar independently of structured tracing.
- Captures exact kernel registrations, normalized `DebugHandle` records, ordered OpSpec bindings, and the reachable upstream Inductor provenance projection.
- Uses deterministic merge-not-replace publication and same-directory atomic writes, with thread-safe in-process publication.
- Preserves richer records across graph breaks, recompiles, cache replay, differing provenance settings, partial failures, and repeated publication.
- Provides a compiler-independent API and CLI for resolving saved profiler events offline.
- Keeps provenance supplemental: collection and publication failures retain bounded diagnostics without replacing compilation failures.

#### Which issue(s) this PR is related to:

Fixes #2578

#### Special notes for your reviewer:

Dependency: #3354 must merge first.

The review amendments do not change the frozen v1 wire shape. They make equivalent contributions merge monotonically across provenance settings, define deterministic unavailable-status precedence, and remove internal project terminology from the shipped contract.

Runtime events currently resolve to the complete finalized bundle. Exact backend command-step-to-OpSpec subsets are unavailable at the open-source boundary and are not inferred.

The real-device small-model event-to-sidecar-to-source path passes end to end.

#### Does this PR introduce a user-facing change?

Yes. Compiled Spyre workloads publish `spyre_provenance.json` under the run-scoped Inductor debug directory by default. Publication can be redirected or disabled with `TORCH_SPYRE_PROVENANCE_PATH`, and saved events can be resolved through `torch_spyre.provenance` without compiler imports or a live compilation cache.

#### Additional note:

Validation:

- Focused provenance gate: 100 passed, including 7 real-device tests
- `tests/test_prepare_kernel.py`: 36 passed
- `tests/profiler/test_ffdc.py`: 25 passed
- Native Kineto provenance trace test: passed
- Pre-commit on affected files and `git diff --check`: passed
- All commits are DCO-signed

### CODE REVIEW

File: torch_spyre/_inductor/provenance_writer.py [hunk 1]
```diff
@@ -0,0 +1,1302 @@
...
+    """Merge and atomically publish one wrapper's collection outcome."""
+    path = resolve_provenance_artifact_path(configured_path)
+    if path is None:
+        return "disabled"
+
+    contribution = _build_contribution(
+        collection, upstream_projection_failed, upstream_projection
+    )
+    result: PublicationResult = "written"
+    with _PUBLICATION_LOCK:
```
Comment (kaoutar55) [inline torch_spyre/_inductor/provenance_writer.py:272]: This lock and the read-merge-write below serialize only threads within one process. With no OS-level file lock, two processes pointed at the same sidecar through `TORCH_SPYRE_PROVENANCE_PATH` can lose an update: both read generation N, both merge, both replace, and the second wins. Either add an `flock`-based critical section around read-merge-write, or correct the "concurrent processes" claim in the PR body and docs. The default PID-scoped path is not affected.

File: torch_spyre/_inductor/provenance_writer.py [hunk 2]
```diff
@@ -0,0 +1,1302 @@
...
+            mode="w",
+            encoding="utf-8",
+            dir=path.parent,
+            prefix=f".{path.name}.",
+            suffix=".tmp",
+            delete=False,
+        ) as temporary:
+            temporary_path = Path(temporary.name)
+            temporary.write(payload)
+        os.replace(temporary_path, path)
```
Comment (kaoutar55) [inline torch_spyre/_inductor/provenance_writer.py:1267]: The temp file is closed and then replaced with no `fsync` in between. The rename is atomic, but on power loss some filesystems commit the rename while the data blocks are still unflushed, which leaves a zero-length sidecar. If durability matters here, add `flush()` and `os.fsync(fileno())` before the rename, and optionally fsync the parent directory. Otherwise a one-line note recording the trade-off is enough.

File: torch_spyre/_inductor/provenance_writer.py [hunk 3]
```diff
@@ -0,0 +1,1302 @@
...
+    if not isinstance(value, dict):
+        raise ProvenanceArtifactError(f"invalid document in {path.name}")
+    return value
+
+
+def _atomic_write(path: Path, payload: str) -> None:
+    temporary_path: Path | None = None
+    try:
+        path.parent.mkdir(parents=True, exist_ok=True)
+        with tempfile.NamedTemporaryFile(
```
Comment (kaoutar55) [inline torch_spyre/_inductor/provenance_writer.py:1255]: `NamedTemporaryFile(delete=False)` leaves a `.spyre_provenance.json.*.tmp` file behind if the process is killed between creation and replace; the `finally` unlink only covers `OSError`. `.gitignore` ignores the literal filename but not this temp pattern. Please add a glob such as `.spyre_provenance.json.*.tmp`.

File: torch_spyre/provenance.py [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+def resolve_provenance_event(
+    event_name: str,
+    artifact_path: str | Path,
+    *,
+    occurrence_selector: str | None = None,
+) -> dict[str, Any]:
+    """Resolve one saved event name against one saved provenance sidecar."""
+    path = Path(artifact_path)
+    try:
+        document = json.loads(path.read_text(encoding="utf-8"))
```
Comment (kaoutar55) [inline torch_spyre/provenance.py]: This path catches `OSError`/`UnicodeError`/`JSONDecodeError` but not `RecursionError`, which both `json.loads` and the recursive `_validate_schema` can raise on a deeply nested sidecar. Since this reader is meant to open sidecars from elsewhere, a crafted file crashes with a traceback instead of a clean diagnostic. Catch `RecursionError` (and broadly `ValueError`) and map it to the diagnostic vocabulary.

File: torch_spyre/provenance.py [hunk 5]
```diff
@@ -0,0 +1,791 @@
...
+    path = (
+        Path(__file__).parent
+        / "_inductor"
+        / "schemas"
+        / "spyre_provenance_v1.schema.json"
+    )
+    return json.loads(path.read_text(encoding="utf-8"))
+
+
+def _validate_schema(
```
Comment (kaoutar55) [inline torch_spyre/provenance.py:421]: This hand-rolled validator, which correctly avoids a runtime `jsonschema` dependency, only supports the keywords the schema uses today. A future schema keyword becomes a silent no-op here while the `jsonschema`-based tests still pass. Please add a test that runs the negative fixtures through this runtime reader, not only through `jsonschema`, so drift is caught.

**Review by kaoutar55 — CHANGES_REQUESTED** [review]
The design is disciplined: deterministic digests, a dependency-free runtime reader, same-directory atomic replace, and good failure isolation. I verified that `jsonschema` stays a test-only dependency, that no timestamps or absolute paths are written into the document, and that the shipped fixtures validate against the schema.

I am requesting changes on one item where the behavior does not match the description, plus some robustness suggestions.

The PR says "concurrent processes writing the same sidecar" with merge-not-replace. That guarantee does not hold. The only lock is a `threading.Lock`, which serializes threads inside one process. There is no OS-level file lock, so two processes can each read generation N, merge, and `os.replace`, and the second write silently drops the first contribution. The rename is atomic so the file never corrupts, but the update is lost. The default path is safe because `_get_debug_dir()` embeds a timestamp and the PID, so each process writes its own file. The race only shows up when `TORCH_SPYRE_PROVENANCE_PATH` points several processes at one shared path, which is exactly the aggregation case the merge logic is built for. Please either add an `flock` critical section around read-merge-write, or correct the claim here and in the docs. The single concurrency test uses a thread pool, so it exercises the thread lock, not the cross-process path.

File: torch_spyre/_inductor/provenance_writer.py (hunk 1 above)
Comment (gsmoon97) [reply torch_spyre/_inductor/provenance_writer.py:278]: Good point. I added an exclusive OS-level lock using `fcntl.flock` around the full read-merge-write transaction. I also added a multiprocess regression test that forces overlapping publications to the same path.

File: torch_spyre/_inductor/provenance_writer.py (hunk 2 above)
Comment (gsmoon97) [reply torch_spyre/_inductor/provenance_writer.py:1281]: I added `flush()` and `fsync()` before replacing the sidecar to make sure the temporary file contents are durable. I left the parent directory unsynchronized because that only affects persistence of the rename across an abrupt system failure. Instead, I documented this limitation in the [PyTorch profiler guide](https://github.com/torch-spyre/torch-spyre/pull/3449/changes#diff-4d0735ea91a69a074dd907128e8bb1b4900956d4fbfd0714fc5ba138bee5fa26L72-R160).

File: torch_spyre/_inductor/provenance_writer.py (hunk 3 above)
Comment (gsmoon97) [reply torch_spyre/_inductor/provenance_writer.py:1269]: Valid point. I added `.spyre_provenance.json.*.tmp` to `.gitignore`, so any temporary files left behind by an abruptly terminated process are ignored.

File: torch_spyre/provenance.py (hunk 4 above)
Comment (gsmoon97) [reply torch_spyre/provenance.py]: `RecursionError` and `ValueError` are now caught during JSON decoding and schema validation. It also returns the existing schema-validation-failure diagnostic. I added focused regressions for both paths.

File: torch_spyre/provenance.py (hunk 5 above)
Comment (gsmoon97) [reply torch_spyre/provenance.py:428]: Added `test_reader_diagnostics_follow_frozen_fixture_manifest` regression that runs every invalid manifest fixture through the dependency-free runtime reader and checks its expected diagnostic. This should catch drift between the JSONSchema and the hand-written validator.

**Review by ashokponkumar — DISMISSED** [review]

**Review by kaoutar55 — DISMISSED** [review]

**Review by dgrove-oss — CHANGES_REQUESTED** [review]
Summary

Adds a compiler-independent spyre_provenance.json v1 sidecar, published alongside compilation (layered on #3354's kernel-identity work) and readable offline via torch_spyre.provenance without any compiler imports. Uses atomic same-directory writes and rank-based monotonic merging so richer provenance records are never clobbered by thinner ones across recompiles/cache-replay/graph-breaks.

Review

BLOCKERs

- [ ] Merge conflict against main. mergeable: CONFLICTING, mergeStateStatus: DIRTY — must be rebased/resolved before merge.
- [ ] CI is fully red (108 fail / 12 pass), and the failure is a real ABI break, not just the conflict. The wheel build/install succeeds, but every test job then fails at import with undefined symbol: _ZN4flex15createDmaParamsE... from libspyre_comms.so.1, surfacing as RuntimeError: Failed to load the backend extension: torch_spyre. The branch has since fallen behind main by two commits (af038a0, b4ae70f); this may resolve itself once rebased, but that needs to be explicitly re-verified with a clean CI run rather than assumed.
- [ ] Unguarded provenance call in torch_spyre/execution/kernel_runner.py, SpyreSDSCKernelRunner.__init__ — calls format_kernel_provenance_event_name(...)/register_kernel_provenance(...) with no try/except, unlike every other provenance call site in this PR (async_compile.py, scheduler.py), which wrap provenance calls specifically so "provenance must never fail the build." A descriptor or _C binding error here would propagate and break kernel construction — directly contradicting the PR's own stated guarantee that provenance failures stay bounded and never replace compilation failures.

SUGGESTIONs

- [ ] torch_spyre/provenance.py hand-rolls a JSON-Schema-subset validator for production use, while tests validate against the real jsonschema library as an independent oracle. The two must be kept in sync by hand — a future schema change could pass the hand-rolled check while violating real JSON Schema semantics undetected. Reasonable to avoid a runtime dependency, but worth a comment flagging the parity requirement.
- [ ] docs/source/user_guide/profiling/pytorch_profiler.md never names TORCH_SPYRE_PROVENANCE_PATH by name even though it's correctly implemented (torch_spyre/_inductor/config.py:4493) — users can't discover it from the docs.
- [ ] The cross-process lock (fcntl.flock in provenance_writer.py) is advisory-only, as all flock is — not a defect, just worth a one-line docstring caveat that a non-cooperating writer isn't protected.

QUESTIONs

- [ ] Can the author confirm whether the .so symbol-resolution CI failure clears once rebased onto current main, or whether it's a separate native/runner-image version skew needing its own fix?
- [ ] _read_existing_document raises ProvenanceArtifactError (doesn't swallow) on a torn/corrupt existing sidecar. Is this always caught as non-fatal at the async_compile.py publish call site, consistent with the PR's claimed tolerance for partial failures? The scenario sweep found a general partial-failure test but not one specifically for corrupt-JSON recovery.

Verdict

REQUEST_CHANGES — resolve the merge conflict, get CI green (confirm the .so symbol error clears, or fix it if not), and wrap the kernel_runner.py provenance call in the same fail-open pattern used everywhere else in this PR. The design itself (atomic writes, monotonic rank-based merge, clean compiler-independent import graph, real jsonschema-backed golden tests) is sound and thoroughly tested — this is close, not a design-level rejection.

**Comment by gsmoon97** [comment]
Thanks @dgrove-oss for the detailed review.

The branch is now restacked onto the latest main. As you suggested, I also made runtime provenance attachment fail-open and added the requested comments about validator parity, `TORCH_SPYRE_PROVENANCE_PATH`, and advisory locking.

Also, to address your questions:

> Can the author confirm whether the .so symbol-resolution CI failure clears once rebased onto current main, or whether it's a separate native/runner-image version skew needing its own fix?

Yes, after rebasing, the full CI passed and the undefined-symbol error did not recur.

> _read_existing_document raises ProvenanceArtifactError (doesn't swallow) on a torn/corrupt existing sidecar. Is this always caught as non-fatal at the async_compile.py publish call site, consistent with the PR's claimed tolerance for partial failures? The scenario sweep found a general partial-failure test but not one specifically for corrupt-JSON recovery.

Yes. `SpyreAsyncCompile.wait()` catches the error, logs and counts it, and allows compilation to continue. I also added a regression using genuinely corrupt JSON that verifies the runtime descriptor survives and the existing file remains untouched.

Please take another look and feel free to ask me any questions. Thanks again!

File: torch_spyre/provenance.py [hunk 6]
```diff
@@ -0,0 +1,796 @@
...
+        reference = schema["$ref"]
+        if not isinstance(reference, str) or not reference.startswith("#/"):
+            raise ValueError(f"{path}: unsupported schema reference")
+        target: object = root
+        for token in reference[2:].split("/"):
+            if not isinstance(target, Mapping):
+                raise ValueError(f"{path}: invalid schema reference")
+            target = target[token.replace("~1", "/").replace("~0", "~")]
+        assert isinstance(target, Mapping)
+        _validate_schema(value, target, root, path)
```
[the remark this answers is not in the document]
Comment (gsmoon97) [reply torch_spyre/provenance.py:448]: Malformed $ref targets now raise a structured schema-validation-failure instead of raising an `AssertionError`.

File: torch_spyre/provenance.py [hunk 7]
```diff
@@ -0,0 +1,796 @@
...
+        raise _ReaderError("schema-validation-failure", "artifact must be an object")
+    if document.get("schemaVersion") != _SCHEMA_VERSION:
+        raise _ReaderError(
+            "unsupported-schema-version",
+            "artifact schema version is not supported",
+        )
+    try:
+        _validate_schema(document, _schema(), _schema(), "$")
+    except (RecursionError, ValueError) as error:
+        raise _ReaderError("schema-validation-failure", str(error)) from None
```
[the remark this answers is not in the document]
Comment (gsmoon97) [reply torch_spyre/provenance.py:403]: Missing/unreadable or invalid packaged schemas now return a structured schema-validation-failure.

File: torch_spyre/_inductor/provenance_writer.py [hunk 8]
```diff
@@ -0,0 +1,1333 @@
...
+
+"""Deterministic validation, merge, and atomic Spyre sidecar publication."""
+
+from __future__ import annotations
+
+import contextlib
+import copy
+import dataclasses
+import fcntl
+import hashlib
```
[the remark this answers is not in the document]
Comment (gsmoon97) [reply torch_spyre/_inductor/provenance_writer.py:22]: `fcntl` is now imported optionally, so its absence no longer prevents the module from importing.

File: torch_spyre/_inductor/provenance_writer.py [hunk 9]
```diff
@@ -0,0 +1,1333 @@
...
+    try:
+        path.parent.mkdir(parents=True, exist_ok=True)
+        directory_fd = os.open(path.parent, os.O_RDONLY)
+        try:
+            fcntl.flock(directory_fd, fcntl.LOCK_EX)
+        except OSError:
+            os.close(directory_fd)
+            raise
+    except OSError:
+        raise ProvenanceArtifactError(f"failed to lock {path.name}") from None
```
[the remark this answers is not in the document]
Comment (gsmoon97) [reply torch_spyre/_inductor/provenance_writer.py:1254]: Unavailable interprocess locking now stops sidecar publication before writing, while the existing fail-open guard logs the failure and lets compilation continue.

**Comment by gsmoon97** [comment]
Hi @dgrove-oss,

I have addressed all the additional review points as well. The PR is rebased to the lastest upstream/main and is ready for final review.

Thanks!

**Review by dgrove-oss — DISMISSED** [review]
Summary

Adds a durable, versioned spyre_provenance.json sidecar mapping compiled Spyre kernels to source-level provenance (debug handles, kernel identities, ordered OpSpec bindings, and the reachable upstream Inductor provenance projection), published via atomic same-directory writes with deterministic merge-not-replace semantics, plus a compiler-independent offline reader/CLI (torch_spyre.provenance). This is a large PR (6674 additions across 25 files) that has already been through multiple review rounds, including my own prior REQUEST_CHANGES.

Review

BLOCKERs

None. All previously-flagged blockers are resolved:
- The unguarded provenance call in kernel_runner.py.__init__ is now wrapped in try/except with fail-open semantics (kernel_runner.py:6055 in the diff).
- The merge conflict / red-CI state is gone — PR is now MERGEABLE, all CI checks pass.
- The cross-process merge race flagged by kaoutar55 (only a threading.Lock, no OS-level lock) is fixed: provenance_writer.py now holds an fcntl.flock(LOCK_EX) on the parent directory fd across the entire read-merge-write critical section, backed by real multiprocessing.get_context("spawn") tests, not just thread-pool tests.

Every provenance call site on the hot compile/execution path (async_compile.py's collection, upstream-projection capture, and publication; kernel_runner.py's runtime attachment; scheduler.py's registration) is wrapped in except Exception with count-once-and-summarize logging — consistent with the PR's "provenance must never fail the build" guarantee.

SUGGESTIONs

- provenance_writer.py (_interprocess_publication_lock): the lock is scoped to the parent directory fd, not a dedicated lock file. Advisory and same-filesystem-only as documented, but any unrelated future feature that also flocks that debug directory would unintentionally serialize against provenance publication. A dedicated .spyre_provenance.json.lock file would scope this more precisely.
- Same function: fcntl is POSIX-only; on Windows this raises and is caught fail-open, but users there get provenance silently and permanently disabled with only a one-time warning. Confirm Windows is genuinely out of scope for Spyre.
- async_compile.py wait(): the _provenance_wait_completed guard makes re-entrant wait() calls silent no-ops after the first. Likely intentional (avoids double-counting / stale V.graph state) but worth a one-line comment explaining why.
- torch_spyre/provenance.py's hand-rolled JSON-Schema-subset validator (used at runtime to avoid a jsonschema prod dependency) must be kept in sync by hand with the real jsonschema-validated test fixtures — a future schema change could pass the hand-rolled check while violating real JSON Schema semantics. Worth a comment flagging the parity requirement.

QUESTIONs

- provenance_writer.py's capture_upstream_projectio.config.trace.provenance_tracking_level, an upstream Inductor-wide config rather than a Spyre-specific knob. Is piggybacking on that global flag intentional, given a user could
enable it for unrelated reasons and inadvertently tpture as a side effect?

Verdict

APPROVE

The design holds up well under scrutiny: atomic samync + os.replace, genuine advisory cross-processlocking validated with real subprocess tests, a merge algorithm that provably never discards richer data (rareplacement with an explicit drift guard), and failll site that touches the compile/execution hot path. License headers, the import regex convention, and doc-to-code consistency (including the previously-missing TORCH_SPYRE_PROVENANCE_PATH mention) all check out. hardening/clarity suggestions, not blockers.

**Comment by gsmoon97** [comment]
@dgrove-oss Yes, this is intentional. The upstream flag only controls whether we can include Inductor’s existing provenance mappings. Spyre sidecar publication is controlled separately by `TORCH_SPYRE_PROVENANCE_PATH`.

The PR has been updated since your approval. Could you please take another look? Thanks!

**Review by dgrove-oss — DISMISSED** [review]
Summary

Adds a durable, versioned spyre_provenance.json sidecar that maps Spyre kernel-provenance keys to normalized debug handles, kernel identities, and occurrences, plus a standalone offline resolver (torch_spyre.provenance, with a CLI). Publication uses deterministic content-addressed IDs, merge-not-replace semantics with a rank ordering that never lets richer data be overwritten by weaker data, and advisory interprocess locking for atomic writes. Provenance collection/publication is layered so failures at any stage (codegen registration, artifact collection, publication, runtime attachment) never abort compilation or execution.

Review

BLOCKERs

None.

SUGGESTIONs

- [ ] The reader (torch_spyre/provenance.py) reimplements a dependency-free JSON Schema interpreter (_validate_schema) and a second semantic validator (_validate_semantics), duplicating logic already present in the writer's validate_provenance_document. This is deliberately justified in comments and is cross-checked against the real jsonschema library and fixtures in tests (test_schema_event_contract_matches_production_codec, test_valid_fixtures_pass_schema_and_semantic_validation), so it's not a blocker — but it's worth flagging as an ongoing maintenance cost: any future schema change must be applied in three places (schema JSON, writer validator, reader validator) and the PR's own fixture-driven tests are the only thing keeping them honest. Consider a code comment at the top of _validate_schema pointing back at the schema file and writer validator so future editors know all three must move together (some of this exists already in test/fixture docs, but not in provenance.py itself).
- [ ] provenance_writer.py and provenance_artifact.py are large (1344 and 515 lines respectively) for new files in one PR. Given the extensive and well-organized test coverage this isn't a practical problem, but future changes to merge/validation logic will have a high review cost per diff.

QUESTIONs

- [ ] The reserved-but-unused selector field on kernelOccurrence in the schema — is this a forward-compat placeholder for a planned follow-up, and if so is there a tracking issue/RFC reference worth linking in the schema's description or the PR body?
- [ ] TORCH_SPYRE_PROVENANCE_PATH="" disables the sidecar entirely (provenance_artifact_path = None). Is there a supported way to detect "provenance was silently disabled" from the artifact side other than reading logs (_log_publication_disabled_once)? Might be worth a one-line doc callout for anyone debugging a "missing sidecar" report.

Verdict

APPROVE

**Comment by gsmoon97** [comment]
Rebased onto the latest upstream/main and resolved all merge conflicts.

Also, please note that the SDSC logging failure is unrelated to this PR, so I did not make any separate test fix in this PR. CI will remain red until upstream fixes those tests.

**Review by dgrove-oss — APPROVED** [review]
lgtm
